TLC5957.cpp

```diff
@@ -0,0 +1,201 @@
+#include "TLC5957.h"
+#include <SPI.h>
+#include <Arduino.h>
+
+void TLC5957::init(uint8_t lat, uint8_t sin, uint8_t sclk, uint8_t gsclk)
+{
+    _lat = lat;
+    _sin = sin;
+    _sclk = sclk;
+    _gsclk = gsclk;
+
+    pinMode(_lat, OUTPUT);
+    pinMode(_sin, OUTPUT);
+    pinMode(_sclk, OUTPUT);
+    pinMode(_gsclk, OUTPUT);
+
+    mSettings = SPISettings(_spi_baud_rate, MSBFIRST, SPI_MODE0);
+    SPI.setMOSI(_sin);
+    SPI.begin();
+
+    digitalWrite(_lat, LOW);
+    setGsclkFrequency(_gsclk_frequency);
+
+}
+
+void TLC5957::setSpiBaudRate(uint32_t baud_rate)
+{
+    _spi_baud_rate = baud_rate;
+    mSettings = SPISettings(_spi_baud_rate, MSBFIRST, SPI_MODE0);
+}
+
+void TLC5957::setGsclkFrequency(uint32_t gsclk_frequency)
+{
+    _gsclk_frequency = gsclk_frequency;
+    analogWriteFrequency(_gsclk, _gsclk_frequency);
+    analogWriteResolution(1);
+    analogWrite(_gsclk, 1);
+}
+
+void TLC5957::latch(int num_edges)
+{
+    // TODO: do I  need to enforce integer division??
+    int lat_delay_microseconds = _spi_baud_rate / 1'000'000;
+    delayMicroseconds(lat_delay_microseconds);
+    digitalWrite(_lat, HIGH);
+    delayMicroseconds(lat_delay_microseconds * num_edges);
+    digitalWrite(_lat, LOW);
+    delayMicroseconds(lat_delay_microseconds);
+}
+
+void TLC5957::setBuffer(uint8_t bit)
+{
+    bitWrite(_buffer, _buffer_count--, bit);
+    SPI.beginTransaction(mSettings);
+    if (_buffer_count == -1)
+    {
+        SPI.transfer(_buffer);
+        _buffer_count = 7;
+        _buffer = 0;
+    }
+    SPI.endTranscation();
+}
+
+void TLC5957::setLodDetection(bool bit_0, bool bit_1)
+{
+    _function_data |= bit_0;
```

Comment:
This actually isn't quite correct. It will set things to 1 but it will never "clear".

Typically for code like this, you want to make sure you have masks shifts defined.

For example, take a look at
https://github.com/PaulStoffregen/cores/blob/master/teensy3/kinetis.h#L875

And how you would use code like this
https://github.com/ramonaoptics/TLC5955/pull/17/files#diff-a50aef481ef64c82473a4792c0d23977d2c64d692feb81808538a2d0eeed387bR470